paddle/fluid/pybind/ir.cc

```diff
@@ -0,0 +1,112 @@
+using paddle::framework::ProgramDesc;
+using paddle::framework::VarDesc;
+using pybind11::return_value_policy;
+
+namespace paddle {
+namespace pybind {
+std::vector<Node *> GetNodesByType(const Node::Type &type, const Graph &graph);
+std::unordered_map<std::string, Node *> GetNodesMapByType(
+    const Node::Type &type, const Graph &graph);
+int Index(const Graph &graph, const Node &op);
+Node *Var(const Graph &graph, const std::string &name);
+Node *Var(const Graph &graph, const std::string &name);
+Node *CreateVar(Graph &graph, const std::string &name, const Node::Type &type,
+                const std::vector<int64_t> &shape);
+bool RemoveVar(Graph &graph, const std::string &name);
+bool RemoveOp(Graph &graph, int index);
+std::shared_ptr<Graph> Clone(Graph &graph);
+
+void BindGraph(py::module *m) {
+  py::class_<Graph, std::shared_ptr<Graph>>(*m, "Graph", "")
+      .def(py::init<const ProgramDesc &>())
+      .def("has", &Graph::Has)
+      .def("get_int", &Graph::Get<int>)
+      .def("get_float", &Graph::Get<float>)
+      .def("get_double", &Graph::Get<double>)
+      .def("get_string", &Graph::Get<std::string>)
+      .def("set", [](Graph &self, const std::string &attr_name,
+                     int attr) { return self.Set(attr_name, new int(attr)); })
+      .def("set",
+           [](Graph &self, const std::string &attr_name,
+              const std::string &attr) {
+             return self.Set(attr_name, new std::string(attr));
+           })
+      .def("set",
+           [](Graph &self, const std::string &attr_name, float attr) {
+             return self.Set(attr_name, new float(attr));
+           })
+      .def("set",
+           [](Graph &self, const std::string &attr_name, double attr) {
+             return self.Set(attr_name, new double(attr));
+           })
+      .def("erase", &Graph::Erase)
+      .def("nodes", &Graph::Nodes, return_value_policy::reference)
+      .def("create_var_node",
+           [](Graph &self, VarDesc &var_desc) -> std::shared_ptr<Node> {
+             return std::shared_ptr<Node>(self.CreateVarNode(&var_desc));
+           })
+      .def("create_op_node",
+           [](Graph &self, OpDesc &op_desc) -> std::shared_ptr<Node> {
+             return std::shared_ptr<Node>(self.CreateOpNode(&op_desc));
+           })
+      .def("create_control_dep_var", &Graph::CreateControlDepVar,
+           return_value_policy::reference)
+      .def("create_empty_node", &Graph::CreateEmptyNode,
+           return_value_policy::reference)
+      .def("release_nodes", &Graph::ReleaseNodes)
+      .def("remove_node",
+           [](Graph &self, Node &node) { return self.RemoveNode(&node); })
+      .def("retrieve_node", &Graph::RetrieveNode,
+           return_value_policy::reference)
+      .def("resolve_hazard", &Graph::ResolveHazard);
+}
+
+void BindNode(py::module *m) {
+  using paddle::framework::ir::CreateNodeForPybind;
+
+  py::class_<Node>(*m, "Node")
+      .def(py::init(&CreateNodeForPybind))
```

Comment:
Unless for test, node should only be created by a graph. There shouldn't be node that belongs to no graph.

Reply:
fixed